### PR TITLE
Adding new test to verify that slots with only zero lamport accounts are saved and recovered correctly

### PR DESCRIPTION
**Pull request by roryharr**
#### Problem
Currently no test verifies that zero lamport accounts are preserved when all other references are older than the snapshot.  If these zero lamport accounts are dropped improperly (either with outstanding references or without preserving dead accounts) it leads to a database failure in rebuild. 

#### Summary of Changes
Confirmed that no other test is exercising this behaviour as this test is failing with the dead account code while existing tests are passing. 

Test should be applicable to the existing behaviour (non dead account code). 
Fixes #

### CODE REVIEW

**Comment by roryharr** [comment]
@brooksprumo Can you suggest some other reviewers for this?

**Comment by brooksprumo** [comment]
> @brooksprumo Can you suggest some other reviewers for this?

Why does it need other reviewers?

**Comment by brooksprumo** [comment]
Here's the current problem statement: 

> Currently no test verifies that slots with only zero lamport accounts are currently preserved across snapshot save and recovery

We do have tests for handling zero lamport accounts. Can you describe what is different when a storage has only zero lamport accounts?

**Comment by roryharr** [comment]
> > @brooksprumo Can you suggest some other reviewers for this?
> 
> Why does it need other reviewers?

I just had no idea if other reviewers are needed for this file or not, given that it is outside of accounts-db

**Comment by brooksprumo** [comment]
> I just had no idea if other reviewers are needed for this file or not, given that it is outside of accounts-db

Makes sense! We talked offline; I'm the perfect person for reviewing.

File: runtime/src/snapshot_bank_utils.rs [hunk 1]
```diff
@@ -2184,6 +2184,168 @@ mod tests {
...
+            BankTestConfig::default(),
+            vec![accounts_dir.clone()],
+        )
+        .wrap_with_bank_forks_for_tests();
+        bank0
+            .transfer(lamports_to_transfer, &mint_keypair, &key2.pubkey())
+            .unwrap();
+        while !bank0.is_complete() {
+            bank0.register_unique_tick();
+        }
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: These can be replaced with `Bank::fill_bank_with_ticks_for_tests()`.

And may only need to be called on the banks that have snapshots taken?

File: runtime/src/snapshot_bank_utils.rs [hunk 2]
```diff
@@ -2184,6 +2184,168 @@ mod tests {
...
+    fn test_snapshots_handle_zero_lamport_accounts() {
+        let collector = Pubkey::new_unique();
+        let key1 = Keypair::new();
+        let key2 = Keypair::new();
+        let key3 = Keypair::new();
+
+        let (_tmp_dir, accounts_dir) = create_tmp_accounts_dir_for_tests();
+        let bank_snapshots_dir = tempfile::TempDir::new().unwrap();
+        let full_snapshot_archives_dir = tempfile::TempDir::new().unwrap();
+        let incremental_snapshot_archives_dir = tempfile::TempDir::new().unwrap();
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: nit: Can be removed.

```suggestion
```

File: runtime/src/snapshot_bank_utils.rs [hunk 3]
```diff
@@ -2184,6 +2184,168 @@ mod tests {
...
+            .transfer(lamports_to_transfer, &key2, &key1.pubkey())
+            .unwrap();
+        bank1
+            .transfer(lamports_to_transfer, &mint_keypair, &key3.pubkey())
+            .unwrap();
+        while !bank1.is_complete() {
+            bank1.register_unique_tick();
+        }
+
+        let _full_snapshot_slot = slot;
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: nit: Remove dead code.

```suggestion
```

File: runtime/src/snapshot_bank_utils.rs [hunk 4]
```diff
@@ -2184,6 +2184,168 @@ mod tests {
...
+        ));
+
+        let fee = bank2.get_fee_for_message(tx.message()).unwrap();
+        let tx = system_transaction::transfer(
+            &key1,
+            &key2.pubkey(),
+            lamports_to_transfer - fee,
+            blockhash,
+        );
+        bank2.process_transaction(&tx).unwrap();
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: Can `Bank::transfer()` be used here instead?

File: runtime/src/snapshot_bank_utils.rs [hunk 5]
```diff
@@ -2184,6 +2184,168 @@ mod tests {
...
+            None,
+            false,
+            false,
+            false,
+            false,
+            Some(ACCOUNTS_DB_CONFIG_FOR_TESTING),
+            None,
+            Arc::default(),
+        )
+        .unwrap_or_else(|err| panic!("Failed to deserialize bank: {:?}", err));
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: I think a `panic!` inside an `or_else` is surprising. The unwrap will output the error already. Is that sufficient?

```suggestion
        .unwrap();
```

Otherwise, could do `.inspect_err()`.

File: runtime/src/snapshot_bank_utils.rs [hunk 6]
```diff
@@ -2184,6 +2184,168 @@ mod tests {
...
+
+        let (_tmp_dir, accounts_dir) = create_tmp_accounts_dir_for_tests();
+        let bank_snapshots_dir = tempfile::TempDir::new().unwrap();
+        let full_snapshot_archives_dir = tempfile::TempDir::new().unwrap();
+        let incremental_snapshot_archives_dir = tempfile::TempDir::new().unwrap();
+        let snapshot_archive_format = ArchiveFormat::Tar;
+
+        let (mut genesis_config, mint_keypair) = create_genesis_config(sol_to_lamports(1_000_000.));
+        // test expects 0 transaction fee
+        genesis_config.fee_rate_governor = solana_sdk::fee_calculator::FeeRateGovernor::new(0, 0);
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: Does this apply to this test? Oh, is it to ensure we know how many lamports to send and have the account end up with a balance of zero?

File: runtime/src/snapshot_bank_utils.rs [hunk 7]
```diff
@@ -2184,6 +2184,168 @@ mod tests {
...
+
+        let lamports_to_transfer = sol_to_lamports(123_456.);
+        let (bank0, bank_forks) = Bank::new_with_paths_for_tests(
+            &genesis_config,
+            Arc::<RuntimeConfig>::default(),
+            BankTestConfig::default(),
+            vec![accounts_dir.clone()],
+        )
+        .wrap_with_bank_forks_for_tests();
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: Is a custom accounts dir needed? I think this can be replaced with `Bank::new_with_bank_forks_for_tests()`.

File: runtime/src/snapshot_bank_utils.rs [hunk 8]
```diff
@@ -2184,6 +2184,168 @@ mod tests {
...
+        // Ensure account1 has been cleaned/purged from everywhere
+        bank3.force_flush_accounts_cache();
+        assert!(
+            bank3.get_account_modified_slot(&key1.pubkey()).is_none(),
+            "Ensure Account1 has been cleaned and purged from AccountsDb"
+        );
+
+        // Take an incremental snapshot and then do a roundtrip on the bank and ensure it
+        // deserializes correctly
+        let full_snapshot_archive_info = bank_to_full_snapshot_archive(
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: This comment does not match the code.

File: runtime/src/snapshot_bank_utils.rs [hunk 9]
```diff
@@ -2184,6 +2184,168 @@ mod tests {
...
+            incremental_snapshot_archives_dir.path(),
+            snapshot_archive_format,
+        )
+        .unwrap();
+
+        // Cleanup the run directory
+        fs::remove_dir_all(&accounts_dir).unwrap();
+
+        let (deserialized_bank, _) = bank_from_snapshot_archives(
+            &[accounts_dir.clone()],
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: I recommend using new/separate account/snapshot dirs for the deserialized bank.

Here's an example:
https://github.com/anza-xyz/agave/blob/0d19c55a3cef99c81312d798040e91c31fc9f1ec/runtime/src/snapshot_bank_utils.rs#L2121-L2125

File: runtime/src/snapshot_bank_utils.rs [hunk 10]
```diff
@@ -2184,6 +2184,168 @@ mod tests {
         assert_eq!(other_incremental_accounts_hash, incremental_accounts_hash);
     }
 
+    /// Test that snapshots are loaded correctly when some slots only contain
+    /// zero lamport accounts
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: I'm not sure the test is really about *slots* with only zero lamport accounts.

File: runtime/src/snapshot_bank_utils.rs [hunk 11]
```diff
@@ -2184,6 +2184,168 @@ mod tests {
...
+        let _full_snapshot_slot = slot;
+        let _full_snapshot_archive_info = bank_to_full_snapshot_archive(
+            bank_snapshots_dir.path(),
+            &bank1,
+            None,
+            full_snapshot_archives_dir.path(),
+            incremental_snapshot_archives_dir.path(),
+            snapshot_archive_format,
+        )
+        .unwrap();
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: What's the point of taking a full snapshot here if we don't use it?

I think it's to have flush/clean/shrink get called? If yes, I recommend we call those functions directly.

File: runtime/src/snapshot_bank_utils.rs [hunk 12]
```diff
@@ -2184,6 +2184,168 @@ mod tests {
...
+        while !bank3.is_complete() {
+            bank3.register_unique_tick();
+        }
+
+        // Ensure account1 has been cleaned/purged from everywhere
+        bank3.force_flush_accounts_cache();
+        assert!(
+            bank3.get_account_modified_slot(&key1.pubkey()).is_none(),
+            "Ensure Account1 has been cleaned and purged from AccountsDb"
+        );
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: Flushing the write cache doesn't clean/purge, so I'm not sure what the intent is here. The assert effectively checks the account is dead (zero lamports).

File: runtime/src/snapshot_bank_utils.rs [hunk 13]
```diff
@@ -2184,6 +2184,168 @@ mod tests {
...
+        .unwrap_or_else(|err| panic!("Failed to deserialize bank: {:?}", err));
+
+        deserialized_bank.wait_for_initial_accounts_hash_verification_completed_for_tests();
+
+        assert!(
+            deserialized_bank
+                .get_account_modified_slot(&key1.pubkey())
+                .is_none(),
+            "Ensure Account1 has not been brought back from the dead"
+        );
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs:2315]: I'd recommend adding a final assert that bank3 equals the deserialized bank.

```rust
assert_eq!(*bank3, deserialized_bank);
```

File: runtime/src/snapshot_bank_utils.rs (hunk 6 above)
Comment (roryharr) [reply runtime/src/snapshot_bank_utils.rs]: Seems like its unneeded, but i'm not sure what he the purpose is. I adapted this test from 
test_incremental_snapshots_handle_zero_lamport_accounts

Removing it from both tests doesn't seem to effect the results of either test.

File: runtime/src/snapshot_bank_utils.rs (hunk 1 above)
Comment (roryharr) [reply runtime/src/snapshot_bank_utils.rs]: Will clean up this pattern in this file.

File: runtime/src/snapshot_bank_utils.rs (hunk 11 above)
Comment (roryharr) [reply runtime/src/snapshot_bank_utils.rs]: The purpose is to capture account1 in the old snapshot. Without this call, Slot1 gets deleted along with slot2 which makes the test pass without exercising the wanted behaviour. 

It looks like flush/squash/clean works. But i'm not really clear what squash does.

Comment (brooksprumo) [reply runtime/src/snapshot_bank_utils.rs]: Gotcha. Can you add a comment in the code that indicates that, please?

File: runtime/src/snapshot_bank_utils.rs (hunk 7 above)
Comment (roryharr) [reply runtime/src/snapshot_bank_utils.rs]: Yup! that works

File: runtime/src/snapshot_bank_utils.rs (hunk 4 above)
Comment (roryharr) [reply runtime/src/snapshot_bank_utils.rs]: Yeah. Is there a better way to get the fee as well?

File: runtime/src/snapshot_bank_utils.rs (hunk 9 above)
Comment (roryharr) [reply runtime/src/snapshot_bank_utils.rs]: Ahh, this is likely why i had to clean up the directories. Will try that method instead.

File: runtime/src/snapshot_bank_utils.rs (hunk 10 above)
Comment (roryharr) [reply runtime/src/snapshot_bank_utils.rs]: I'm trying to figure out the best way to describe this differently than some of the other zero lamport account tests... Any ideas?

Comment (brooksprumo) [reply runtime/src/snapshot_bank_utils.rs]: Keeping it about zero lamport *accounts* is fine, I think the *slots* part should be removed.

How about something like:
> Test that snapshots correctly handle zero lamport accounts

File: runtime/src/snapshot_bank_utils.rs (hunk 4 above)
Comment (brooksprumo) [reply runtime/src/snapshot_bank_utils.rs]: Probably not better. Could log the fee here and then hard code it into the test. Now that I realize it is to ensure we send all the lamports and end up with a balance of zero lamports, I'm fine with the current impl.

File: runtime/src/snapshot_bank_utils.rs [hunk 14]
```diff
@@ -1187,9 +1187,7 @@ mod tests {
         let genesis_config = GenesisConfig::default();
         let original_bank = Bank::new_for_tests(&genesis_config);
 
-        while !original_bank.is_complete() {
-            original_bank.register_unique_tick();
-        }
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: Can you back out all these changes, please? Let's keep this PR focused on adding the new test, and a separate PR can handle changing these over.

File: runtime/src/snapshot_bank_utils.rs (hunk 11 above)
Comment (brooksprumo) [reply runtime/src/snapshot_bank_utils.rs]: I see this snapshot has now been removed. So the snapshot wasn't needed then? It's not a test of latest full snapshot slot? Only needed to ensure flush/clean/shrink ran?

Comment (roryharr) [reply runtime/src/snapshot_bank_utils.rs]: It turns out it needed the squash.

**Review by brooksprumo — APPROVED** [review]
:shipit:
